db/dcrpg/internal/treasury.go

```diff
@@ -43,11 +43,27 @@ const (
 	InsertTreasuryRowOnConflictDoNothing = InsertTreasuryRow + `ON CONFLICT (tx_hash, block_hash)
 		DO NOTHING;`
 
-	SelectTreasuryTxns = `SELECT * FROM treasury WHERE is_mainchain ORDER BY block_height DESC
+	SelectTreasuryTxns = `SELECT * FROM treasury 
+		WHERE is_mainchain
+		ORDER BY block_height DESC
 		LIMIT $1 OFFSET $2;`
 
-	SelectTreasuryBalance = `SELECT COUNT(1), SUM(value) AS balance, SUM(CASE WHEN value<0 THEN -value ELSE 0 END) AS spent
-		FROM treasury WHERE is_mainchain AND block_height <= $1;`
+	SelectTypedTreasuryTxns = `SELECT * FROM treasury 
+		WHERE is_mainchain
+			AND tx_type = $1
+		ORDER BY block_height DESC
+		LIMIT $2 OFFSET $3;`
+
+	SelectTreasuryBalance = `SELECT
+		COUNT(CASE WHEN tx_type=4 THEN 1 END) as tadds,
+		SUM(CASE WHEN tx_type=4 THEN value ELSE 0 END) AS spent,
+		COUNT(CASE WHEN tx_type=5 THEN 1 END) as tspends,
+		SUM(CASE WHEN tx_type=5 THEN -value ELSE 0 END) AS spent,
+		COUNT(CASE WHEN tx_type=6 THEN 1 END) as tbases,
+		SUM(CASE WHEN tx_type=6 THEN value ELSE 0 END) AS spent
+		FROM treasury
+		WHERE is_mainchain AND block_height <= $1;`
```

Comment:
I think I might have gone with a `GROUP BY (tx_type)` for this, but yours does make the scan easier I think, and the cost is really not bad, so I like yours.

For ref `SELECT, count(*), sum(value) FROM treasury WHERE is_mainchain AND block_height <= $1 GROUP BY tx_type;`

Reply:
Where I'm at right now is 

```
SelectTreasuryBalance = `SELECT
	tx_type,
	COUNT(CASE WHEN block_height <= $1 THEN 1 END),
	COUNT(CASE WHEN block_height > $1 THEN 1 END),
	SUM(CASE WHEN block_height <= $1 THEN value ELSE 0 END),
	SUM(CASE WHEN block_height > $1 THEN value ELSE 0 END)
	FROM treasury
	WHERE is_mainchain
	GROUP BY tx_type;`
```

Though looking at this now, I'm wondering if it would be faster to grab do 

```
COUNT(CASE WHEN block_height <= $1 THEN 1 END),
COUNT(1),
SUM(CASE WHEN block_height <= $1 THEN value ELSE 0 END),
SUM(value)
```

and then take the difference in the caller.